File: Local-TFG/Manejo de peticiones/prueba-1/python/worker.py
from pyscript import sync, display, document, fetch
import asyncio
import time
import json

PYTHON_SERVER = "http://127.0.0.1:5001"


async def do_analisis():
    try:
        num_requests = int(document.getElementById(
            "num-requests-pyscript").value)
        delay = int(document.getElementById("request-delay-pyscript").value)

        urls = [
            f"{PYTHON_SERVER}/mock-api/{delay}" for _ in range(num_requests)]

        start_time = time.perf_counter()
        individual_times = []

        async def fetch_url(url):
            t0 = time.perf_counter()
            resp = await fetch(url)
            elapsed = (time.perf_counter() - t0) * 1000

            individual_times.append(elapsed)

            if resp.status == 200:
                try:
                    python_dict = await resp.json()
                    return python_dict
                except Exception as e:
                    return {"error": f"JSON parse failed: {str(e)}"}
            return {"error": f"Request failed with status {resp.status}"}

        results = await asyncio.gather(*(fetch_url(url) for url in urls))

        total_time = (time.perf_counter() - start_time) * 1000
        avg_time = sum(individual_times) / \
            len(individual_times) if individual_times else 0

        result = {
            "average_time_ms": avg_time,
            "total_time_ms": total_time,
            "total_requests": num_requests,
            "last_value": results[-1]["data"][-1]["value"]
        }

        return json.dumps(result)
    except Exception as e:
        display(f"Error: {e}", target="pyscript-output")

sync.do_analisis = do_analisis

File: Local-TFG/Manejo de peticiones/prueba-1/javascript/worker.js
self.onmessage = async function (e) {
    const data = e.data;

    if (data.type === "do_analisis") {
        try {
            const { id, num_requests, delay } = data;
            const JAVASCRIPT_SERVER = "http://127.0.0.1:5002";

            const results = [];
            const individualTimes = [];

            const fetchWithTiming = async () => {
                const url = `${JAVASCRIPT_SERVER}/mock-api/${delay}`;
                const start = performance.now();
                try {
                    const res = await fetch(url);
                    const json = await res.json();
                    results.push(json);
                } catch (e) {
                    results.push(null);
                } finally {
                    const end = performance.now();
                    individualTimes.push(end - start);
                }
            };

            const start_time = performance.now();
            await Promise.all(
                Array.from({ length: num_requests }, () => fetchWithTiming())
            );
            const total_time = performance.now() - start_time;

            const avg_time = individualTimes.length > 0
                ? individualTimes.reduce((acc, t) => acc + t, 0) / individualTimes.length
                : 0;

            let last_value = null;
            for (let i = results.length - 1; i >= 0; i--) {
                const r = results[i];
                if (r && r.data && Array.isArray(r.data)) {
                    last_value = r.data[r.data.length - 1].value;
                    break;
                }
            }

            const result = {
                average_time_ms: avg_time,
                total_time_ms: total_time,
                total_requests: num_requests,
                last_value,
            };

            self.postMessage({ id, json: JSON.stringify(result) });
        } catch (error) {
            self.postMessage({
                id: data.id,
                error: `Error in worker: ${error.message}`,
            });
        }
    }
};
